6_nodejs/week2/app.js

```diff
@@ -0,0 +1,110 @@
+import express from "express";
+import fs from "fs";
+const data = fs.readFileSync("./documents.json");
+let documents = JSON.parse(data);
+//console.log(documents);
+const app = express();
+const port = process.env.PORT || 3000;
+// Support parsing JSON requests
+app.use(express.json());
+// For parsing application/x-www-form-urlencoded
+//app.use(express.urlencoded({ extended: true }));
+app.get("/", (req, res) => {
+  res.send("This is a search engine.");
+});
+//---------------------------------------------------------------
+//GET /search
+/* This endpoint will accept a query parameter called q, short for query. 
+If q is not provided, the endpoint should return all documents.
+If q is provided, the endpoint should return the documents with some field that matches the value of q.
+Example response if we call GET   /search?q=hello :
+[
+    {
+        "id": 2,
+        "type": "doc",
+        "value": "hello world"
+    }
+]
+*/
+app.get("/search", (req, res) => {
+  const query = req.query.q; // Extract the 'q' query parameter from the request URL
+  if (!query) {
+    res.json(documents); // If 'q' is not provided, it returns all documents
+  } else {
+    // filter documents' objects, as doc, based on whether any field matches the query string
+    const filteredDocuments = documents.filter((doc) =>
+      Object.values(doc).some(
+        (value) => typeof value === "string" && value.includes(query)
+      )
+    );
+    if (filteredDocuments) {
+      res.json(filteredDocuments);
+    } else {
+      res.send("no match found");
+    }
+  }
+});
+//----------------------------------------------------------------
+//GET /documents/:id
+//find and respond with the document matching the id parameter.
+//If there is no such document, respond with a 404 Not Found.
+//You can assume that the document IDs are unique so there's no need to handle duplicates.
+app.get("/documents/:id", (req, res) => {
+  console.log(req.params);
+  const documentId = +req.params.id;
```

Comment:
I suppose you have a plus here because `id` in the JSON file is a number. In general ID's are strings, so the plus here would cause you some trouble. In the "real world" an ID is very often something called GUID. You can check out this generator https://www.guidgenerator.com/ . It contains both numbers and letters and it is so long that it guarantees that the combination will always be unique. It is always safer to assume that ID is a string.